// @ledgerhq/hw-transport-node-hid-singleton/lib/listenDevices.js
"use strict";
exports.__esModule = true;
exports.listenDevices = exports.setUsbDebounce = void 0;
var usb_1 = require("usb");
var devices_1 = require("@ledgerhq/devices");
var logs_1 = require("@ledgerhq/logs");
var deviceToLog = function (_a) {
    var idProduct = _a.deviceDescriptor.idProduct, busNumber = _a.busNumber, deviceAddress = _a.deviceAddress;
    return "productId=".concat(idProduct, " busNumber=").concat(busNumber, " deviceAddress=").concat(deviceAddress);
};
var usbDebounce = 1000;
var setUsbDebounce = function (n) {
    usbDebounce = n;
};
exports.setUsbDebounce = setUsbDebounce;
var mapRawDevice = function (_a) {
    var locationId = _a.busNumber, deviceAddress = _a.deviceAddress, _b = _a.deviceDescriptor, vendorId = _b.idVendor, productId = _b.idProduct, serialNumber = _b.iSerialNumber;
    return ({
        locationId: locationId,
        vendorId: vendorId,
        productId: productId,
        deviceName: "",
        manufacturer: "",
        serialNumber: serialNumber,
        deviceAddress: deviceAddress
    });
};
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
var listenDevices = function (onAdd, onRemove) {
    var timeout;
    var add = function (device) {
        if (device.deviceDescriptor.idVendor !== devices_1.ledgerUSBVendorId)
            return;
        (0, logs_1.log)("usb-detection", "add: " + deviceToLog(device));
        if (!timeout) {
            // a time is needed for the device to actually be connectable over HID..
            // we also take this time to not emit the device yet and potentially cancel it if a remove happens.
            timeout = setTimeout(function () {
                onAdd(mapRawDevice(device));
                timeout = null;
            }, usbDebounce);
        }
    };
    var remove = function (device) {
        if (device.deviceDescriptor.idVendor !== devices_1.ledgerUSBVendorId)
            return;
        (0, logs_1.log)("usb-detection", "remove: " + deviceToLog(device));
        if (timeout) {
            clearTimeout(timeout);
            timeout = null;
        }
        else {
            onRemove(mapRawDevice(device));
        }
    };
    usb_1.usb.on("attach", add);
    usb_1.usb.on("detach", remove);
    return function () {
        if (timeout)
            clearTimeout(timeout);
        usb_1.usb.unrefHotplugEvents();
    };
};
exports.listenDevices = listenDevices;
process.on("exit", function () {
    usb_1.usb.unrefHotplugEvents();
});
